src/System.Management.Automation/engine/Utils.cs

```diff
@@ -1,41 +1,265 @@
+                return false;
+            }
+
+            ul = (ulong)value;
+            return true;
+        }
+
+        internal static bool TryCast(BigInteger value, out decimal dm)
+        {
+            if (value < (BigInteger)decimal.MinValue || (BigInteger)decimal.MaxValue < value)
+            {
+                dm = 0;
+                return false;
+            }
+
+            dm = (decimal)value;
+            return true;
+        }
+
+        internal static bool TryCast(BigInteger value, out double db)
+        {
+            if (value < (BigInteger)double.MinValue || (BigInteger)double.MaxValue < value)
+            {
+                db = 0;
+                return false;
+            }
+
+            db = (double)value;
+            return true;
+        }
+
+        /// <summary>
+        /// Parses a given string or ReadOnlySpan&lt;char&gt; to calculate its value as a binary number.
+        /// Assumes input has already been sanitized and only contains zeroes (0) or ones (1).
+        /// </summary>
+        /// <param name="digits">Span or string of binary digits. Assumes all digits are either 1 or 0.</param>
+        /// <param name="unsigned">
+        /// Whether to treat the number as unsigned. When false, respects established conventions
+        /// with sign bits for certain input string lengths.
+        /// </param>
+        /// <returns>Returns the value of the binary string as a BigInteger.</returns>
+        internal static BigInteger ParseBinary(ReadOnlySpan<char> digits, bool unsigned)
+        {
+            if (!unsigned)
+            {
+                if (digits[0] == '0')
+                {
+                    unsigned = true;
+                }
+                else
+                {
+                    switch (digits.Length)
+                    {
+                        // Only accept sign bits at these lengths:
+                        case 8: // byte
+                        case 16: // short
+                        case 32: // int
+                        case 64: // long
+                        case 96: // decimal
+                        case int n when n >= 128: // BigInteger
+                            break;
+                        default:
+                            // If we do not flag these as unsigned, bigint assumes a sign bit for any (8 * n) string length
+                            unsigned = true;
+                            break;
+                    }
+                }
+            }
+
+            // Only use heap allocation for very large numbers
+            const int MaxStackAllocation = 512;
+
+            // Calculate number of 8-bit bytes needed to hold the input,  rounded up to next whole number.
+            int outputByteCount = (digits.Length + 7) / 8;
+            int outputByteIndex = outputByteCount - 1;
+
+            Span<byte> outputBytes = outputByteCount <= MaxStackAllocation ? stackalloc byte[outputByteCount] : new byte[outputByteCount];
```

Comment:
CoreFX team recommend to use a const for stack allocations to get more fast code.
```suggestion
            Span<byte> outputBytes = outputByteCount <= MaxStackAllocation ? stackalloc byte[MaxStackAllocation] : new byte[outputByteCount];
```

Reply:
I did give this a look, but this completely breaks parsing due to it typically filling the bytes from first to last, and the actual BigInteger constructor being used as big-endian. 

This _could_ be worked in, I think, but I think it would require additional allocations that I think it would likely be more effective to avoid.

Reply:
@iSazonov once I got the logic fixed, which wasn't nearly as hard as I first thought, here are the results of two different methods of doing it vs the current method.

The first alternate method slices the over-allocated span down to size before passing it to the BigInteger constructor, and the second simply ensures that all bytes are positioned in the span to be correctly parsed, which just meant redefining the starting point to be the end of the span, instead of counting the bytes needed there.

https://gist.github.com/vexx32/375d5553dfe3edee60cd9616228dd3a5

In this instance, it seems that even allocating according to a `const` value is slower thanks to the extra allocations and/or operations needed somewhere along the line.

Reply:
Using const speeds up _calling_ the method not the method itself. Are you sure that your test measure this correctly?

Reply:
Even if it speeds up calling the method, the excess allocation is taking far more time than it could possibly be saving on the method call, unless method calls are an order of magnitude more expensive than I am generally led to believe.

But be that as it may, I'm not really sure how one would measure that.

Reply:
Sorry I was not accurate. The compiler optimization can be performed only if `stackalloc Const` is not under _a condition_.
I mean a code like:
```suggestion
            Span<byte> outputBytes = stackalloc byte[MaxStackAllocation];
            if (outputByteCount > MaxStackAllocation)
            {
                outputBytes = new byte[outputByteCount];
            }
```
I'm ok with current code too.

Reply:
@iSazonov I'm not sure what kind of perf benefit this _should_ give, but all my attempts to include this suggestion haven't seemed to produce the results you're describing.

https://gist.github.com/vexx32/d2b7fff82f93635ea974bdbc0c26fa89

This is the benchmarking code I'm using, with BenchmarkDotNet targeting `netcoreapp2.1`; if you're able to take a look and see what I might be doing wrong here that'd be awesome.

Thank you for all your fantastic help! 😄